src/OrchardCore/OrchardCore/Extensions/Features/TypeFeatureProvider.cs

```diff
@@ -1,26 +1,55 @@
 using System;
 using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
 using OrchardCore.Environment.Extensions.Features;
 
 namespace OrchardCore.Environment.Extensions
 {
     public class TypeFeatureProvider : ITypeFeatureProvider
     {
-        private readonly ConcurrentDictionary<Type, IFeatureInfo> _features = new();
+        private readonly ConcurrentDictionary<Type, IEnumerable<IFeatureInfo>> _features = new();
+
+        public IExtensionInfo GetExtensionForDependency(Type dependency)
+        {
+            if (_features.TryGetValue(dependency, out var features))
+            {
+                return features.First().Extension;
+            }
+
+            throw new InvalidOperationException($"Could not resolve extension for type {dependency.Name}");
+        }
 
         public IFeatureInfo GetFeatureForDependency(Type dependency)
         {
-            if (_features.TryGetValue(dependency, out var feature))
+            if (_features.TryGetValue(dependency, out var features))
             {
-                return feature;
+                // Gets the first feature that has the same ID as the extension, if any. Otherwise returns the
+                // first feature.
+                return features.FirstOrDefault(feature => feature.Extension.Id == feature.Id) ?? features.First();
             }
 
-            throw new InvalidOperationException($"Could not resolve feature for type {dependency.Name}");
+            throw new InvalidOperationException($"Could not resolve main feature for type {dependency.Name}");
+        }
+
+        public IEnumerable<IFeatureInfo> GetFeaturesForDependency(Type dependency)
+        {
+            if (_features.TryGetValue(dependency, out var features))
+            {
+                return features;
+            }
+
+            throw new InvalidOperationException($"Could not resolve features for type {dependency.Name}");
+        }
+
+        public IEnumerable<Type> GetTypesForFeature(IFeatureInfo feature)
+        {
+            return _features.Where(kv => kv.Value.Contains(feature)).Select(kv => kv.Key);
```

Comment:
materialize this using `.ToArray()` to prevent evaluating this every time

Reply:
Callers are enumerating only once, one even with an `Any()` condition. `ToArray()` would be worse performance wise.